### PR TITLE
Prevent NPE on backwards seek on Chunker

### PR DESCRIPTION
**Pull request by werkt**
Ensure that a reset() is not immediately followed by a data member
access with a call to maybeInitialize(). Practically, this can occur if
an offset has progressed locally, but the remote indicates a non-0
committed size, and would always have thrown an NPE on a requisite seek,
proven with the accompanied test case.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/Chunker.java [hunk 1]
```diff
@@ -142,13 +142,12 @@ public void reset() throws IOException {
    * <p>Closes any open resources (file handles, ...).
    */
   public void seek(long toOffset) throws IOException {
-    maybeInitialize();
+    boolean resetAndSkip = toOffset < offset;
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: that's unused?

File: src/main/java/com/google/devtools/build/lib/remote/Chunker.java [hunk 2]
```diff
@@ -142,13 +142,12 @@ public void reset() throws IOException {
...
     if (toOffset < offset) {
       reset();
-      if (toOffset != 0) {
-        data.skip(toOffset);
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: if you added a maybeInitialize() above this line above data.skip() wouldn't that be enough?

Comment (werkt) [reply src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: Yeah, there's probably too much cleverness going on in this already - I was avoiding the initialize and skip if it was a noop or could be left in the reset condition.

File: src/main/java/com/google/devtools/build/lib/remote/Chunker.java [hunk 3]
```diff
@@ -142,15 +142,11 @@ public void reset() throws IOException {
...
-      }
-    } else if (offset != toOffset) {
-      data.skip(toOffset - offset);
     }
+    maybeInitialize();
+    data.skip(toOffset - offset);
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: While you are at it: I just noticed another bug with this code. We ignore the return value of InputStream#skip. According to its documentation it's free to skip less bytes than specified. Similar to read.

Do you want to fix this here or should we do it in a different change?

Comment (werkt) [reply src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: I'll do it here

Comment (werkt) [reply src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: There are crazy directions to go in with the response from the skip here: more than expected, 0 with a possible non-eof. I'll try to accommodate them all here, and throw if we get into a bad state. lmk if it's too much

Comment (werkt) [reply src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: Nevermind, I'll just switch to using ByteStreams.skipFully, it accomplishes all that I wanted it to.

Comment (buchgr) [reply src/main/java/com/google/devtools/build/lib/remote/Chunker.java]: Today I learned. Beautiful. Thanks!

**Comment by buchgr** [comment]
LGTM. I ll import.

**Comment by laurentlb** [comment]
Is it a recent regression? (did this work with a recent version of Bazel)

**Comment by werkt** [comment]
@laurentlb No, this has been present since 0.26.0, introduced with https://github.com/bazelbuild/bazel/pull/7846

**Comment by laurentlb** [comment]
I don't think it qualifies for a cherry-pick, but it will definitely be part of Bazel 2.2 in February.

**Comment by werkt** [comment]
Just an update here, I'm pretty sure this is a regression in the transition to 2.0. It may simply be emergent, that Jacobs refactoring made everything faster and induced it, but we have only started seeing it with clients using 2, and not 1

**Comment by jlisee** [comment]
> I don't think it qualifies for a cherry-pick, but it will definitely be part of Bazel 2.2 in February.

I represent the bazel users at Uber ATG that leverage @werkt Buildfarm project.  We hit the NPEs often enough that we had to delay our roll out of Bazel 2 by what will eventually be 1.5 to 2 months.

This is feels like an ideal candidate for a cherry pick and patch release, because:
 - It delays users upgrading
 - Is a non-breaking change
 - Fixes an underlying stability issues
 - Comes with an accompanying unit test

If you have another patch release of schedule before 2.2 it would be great to see this included in that release.

**Comment by buchgr** [comment]
I can see that the bug is severe enough that a patch release could be warranted. I am surprised there aren't more bug reports about this.

**Comment by laurentlb** [comment]
Thanks for the feedback, I'll take a look at it.

cc @philwo

**Comment by laurentlb** [comment]
@jlisee, @werkt 
See Bazel 2.1.1rc: https://github.com/bazelbuild/bazel/issues/10132#issuecomment-585747185

I'd interested to hear what we can do to help you roll out new versions of Bazel. I think it would also be beneficial for everyone if you had a simple way to try new versions of Bazel (or even build it at head every day) and report issues more quickly.

We can easily mark bugs as "release critical" when they've just been introduced. Older bugs are usually not blocking.

**Comment by werkt** [comment]
tl;dr: @laurentlb we can't keep up with you, yet.

> I'd interested to hear what we can do to help you roll out new versions of Bazel. I think it would also be beneficial for everyone if you had a simple way to try new versions of Bazel (or even build it at head every day) and report issues more quickly.

This is an extremely difficult to address problem for us. The cadence of bazel is intense, and it is not completely a goal for us to stay up to date with the latest bleeding edge commits there, for a variety of reasons. Even for releases, *every single minor version* since 0.4, including the post-1 minor releases, has required changes to our code or build definitions. Upgrades are major events, not the least of which because they are the most likely places for all remote action keys to change their composition, resulting in novel cache keying.

We use `bazilisk --migrate` internally, but the inconsistency of this with regard to the duration that incompatible flags remain in bazel to the release cadence means that we see red herring incompatibles (that we can only verify are red herrings through release candidate evaluation or commit diving) that don't prevent us from moving to the next version of bazel, but complicate our evaluation of the results.

And even with feature deprecation-based migration, many releases simply come with a newly broken (hyrum-esque) feature, that we find out when we can finally evaluate the release. Non-json profiling has been broken for some time. This bug, caused still by an unknown coincidence of behaviors in the client, broke for our particular (standard-compliant) remote implementation. Performance degradation in a language-specific command generation section of the composition will have caused us to reevaluate overall build timeouts.

We are trying to keep up and attempting to put in place ways to get the latest bazel tested, but sometimes reporting new bugs in a timely fashion loses out over other priorities, and a release slips us for a couple of months. And then we're a major release behind.

This is blatantly selfish, or at least shortsighted, but the best way for us to roll out new versions of bazel is to have a lower frequency of new versions of bazel. Better still, fewer versions with any *known* incompatibilities.
> 
> We can easily mark bugs as "release critical" when they've just been introduced. Older bugs are usually not blocking.

Similar to the above, I don't think we're watching every bug that comes in, mostly due to the onerous filtering that requires beyond just being 'release critical' - we build on a specific OS, with a specific workflow, with a specific feature set, with specific strategies, and with very specific languages and tooling. The firehose is just too much.
